### PR TITLE
LUCENE-9202: refactor leaf collectors in TopFieldCollector

### PR DESCRIPTION
**Pull request by msokolov**
Refactoring the two LeafCollector subclasses here into a shared `TopFieldLeafCollector` class to make the code easier on the eyes and amenable to further complications

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java [hunk 1]
```diff
@@ -555,7 +519,7 @@ public static void populateScores(ScoreDoc[] topDocs, IndexSearcher searcher, Qu
 
   final void add(int slot, int doc) {
     bottom = pq.add(new Entry(slot, docBase + doc));
-    queueFull = totalHits == numHits;
+    queueFull = slot == numHits - 1;
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java]: Hmm why was this needed?  Did we lose (easy) access to `totalHits`?

Can we `assert slot < numHits`?

File: lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java [hunk 2]
```diff
@@ -69,6 +69,79 @@ public void setScorer(Scorable scorer) throws IOException {
     }
   }
 
+  private abstract class TopFieldLeafCollector extends MultiComparatorLeafCollector {
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java]: Should we maybe pull this out into its own `TopFieldLeafCollector.java` source?

Comment (msokolov) [reply lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java]: Well it has many reliances on `TopFieldCollector`'s members, and is never created on its own, so I think it's probably best as an inner class.

Comment (msokolov) [reply lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java]: I was mostly focused on the code duplicattion. However there are other opportunities to simplify - eg we could merge with `MultiComparatorLeafCollector`.

File: lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java (hunk 1 above)
Comment (msokolov) [reply lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java]: It's because in `PagingFieldCollector` we care about the number of hits collected since the first page, rather than `totalHits`, and this abstracts away the difference, since `slot` is either `totalHits-1` or `hitsCollected-1`

Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java]: maybe leave a comment about it there?

File: lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java (hunk 2 above)
Comment (mikemccand) [reply lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java]: OK let's do that in followon issue.  Progress not perfection!

File: lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java (hunk 1 above)
Comment (msokolov) [reply lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java]: Thanks, I did

**Comment by msokolov** [comment]
pushed via command line
